### PR TITLE
Task4 Managing Available Actions

### PR DESCRIPTION
**Pull request by tsmith-dev**
This branch implements the last piece of Task 4, making available actions only available to logged in users. I'm planning to expand this to the other resources in the application, as well, so that only admins can create Projects, Locations, Departments, etc.

### CODE REVIEW

File: test/integration/employees_edit_test.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
+require 'test_helper'
+
+class EmployeesEditTest < ActionDispatch::IntegrationTest
+
+  def setup
+    @employee = employees(:MyString)
+  end
+ 
+  test "successful edit with friendly forwarding" do
```
Comment (v-ramachandran) [inline test/integration/employees_edit_test.rb:9]: You don't want to have large test cases like this, where the actual case itself is lengthy and contains several assertions. How will you know what is broken? Test cases fail upon the first assertion failure. Having multiple assertions in a single test means that you can't really figure out what all is broken at once. In addition, it makes you really contain a lot of logic in your test case which will start expanding the test case's purpose, which can also be problematic.

An alternative practice is to keep test cases short and sweet. Specifically, this means that you try to keep the test case to one assertion as much as possible. This will represent a single item/behavior/functionality that you are trying to test. For more assertions you make, simply create another test case. This way, you can detect as much defect/fault/errors as possible at once. In addition, each of your test cases is short and has a very specific purpose. This adds clarity to your test cases and, done well, allows for your test cases to be read like documentation. 

Consider grouping your individual test cases as well. It will help in terms of organization, DRYing, and overall readability.

Comment (v-ramachandran) [inline test/integration/employees_edit_test.rb:9]: Keep your comments for tests very specific. This is ambiguous. If I were to say something like, "When X happens, Y results," "When X happens, Y has Z items," or "X renders Y page," it is very clear and direct what is actually being tested.

File: test/integration/employees_edit_test.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+    get edit_employee_path(@employee)
+    log_in_as(@employee)
+    assert_redirected_to edit_employee_path(@employee)
+    name  = "Foo Bar"
```
Comment (v-ramachandran) [inline test/integration/employees_edit_test.rb:13]: Why not put this in the setup as well?

File: app/views/sessions/new.html.erb [hunk 3]
```diff
@@ -11,6 +11,11 @@
       <%= f.label :password %>
       <%= f.password_field :password, class: 'form-control' %>
 
+      <%= f.label :remember_me, class: "checkbox inline" do %>
```
Comment (v-ramachandran) [inline app/views/sessions/new.html.erb:14]: For the sake of simplicity, I just want to propose an alternative. Why not always assume 'remember' every single time?

File: app/views/layouts/application.html.erb [hunk 4]
```diff
@@ -7,34 +7,45 @@
...
+                </ul>
+             </li>
+           <% else %>
+             <li><%= link_to "Log in", login_path %></li>
+           <% end %>
+        </ul>
+      </nav>
+    </div>
+  </header>
     <div class="container">
```
Comment (v-ramachandran) [inline app/views/layouts/application.html.erb:48]: Where does this close?

File: app/views/layouts/application.html.erb [hunk 5]
```diff
@@ -7,34 +7,45 @@
   
 </head>
 <body>
-    <div class="container">
-      <% flash.each do |message_type, message| %>
-        <div class="alert alert-<%= message_type %>"><%= message %></div>
-      <% end %>
+  <div class="container">
+    <% flash.each do |message_type, message| %>
+  <div class="alert alert-<%= message_type %>"><%= message %></div>
```
Comment (v-ramachandran) [inline app/views/layouts/application.html.erb:12]: Your indentations here seem off.

File: app/views/employees/_form.html.erb [hunk 6]
```diff
@@ -11,11 +11,11 @@
   </div>
   <div class="field">
     <%= f.label :password %>
-    <%= f.password_field :password, class: 'form-control', :placeholder => 'Enter your password...' %>
+    <%= f.password_field :password, class: 'form-control', :placeholder => 'Enter new password...' %>
```
Comment (v-ramachandran) [inline app/views/employees/_form.html.erb:14]: Why this change? Will it not be confusing if the person tries to sign up for the first time? Or this is explicitly for changing information?

File: app/views/layouts/application.html.erb [hunk 7]
```diff
@@ -7,34 +7,45 @@
   
 </head>
 <body>
-    <div class="container">
-      <% flash.each do |message_type, message| %>
-        <div class="alert alert-<%= message_type %>"><%= message %></div>
-      <% end %>
+  <div class="container">
+    <% flash.each do |message_type, message| %>
```
Comment (v-ramachandran) [inline app/views/layouts/application.html.erb:11]: Be careful doing something like this. While the idea is great, what if someone in the future does something like flash[:foo]="bar"?

Instead, why not only look for a select set of keys from the flash?

File: app/controllers/departments_controller.rb [hunk 8]
```diff
@@ -1,5 +1,6 @@
 class DepartmentsController < ApplicationController
   before_action :set_department, only: [:show, :edit, :update, :destroy]
```
Comment (v-ramachandran) [inline app/controllers/departments_controller.rb:2]: In general, you may be better off not using before actions to achieve setup like this. Before actions let you achieve common things that are not directly related to the business logic of the method. For example, you should be able to directly see all things that are relevant to the logic in the function. Something like login doesn't impact the logic in the controller method functions. However, setting up variables are different. This may lead to frustrating bugs. I would recommend against this especially since your setup here involves only a single variable being setup.

File: app/controllers/departments_controller.rb [hunk 9]
```diff
@@ -1,5 +1,6 @@
...
+  before_action :admin_employee
```
Comment (v-ramachandran) [inline app/controllers/departments_controller.rb:3]: Your actions should read like actions. You want your syntax here to read well. What is the method admin_employee doing? Can that be somewhat inferred without having to dig through the method itself? You can accomplish that by using something like "verify_admin_employee" versus "admin_employee"

File: app/controllers/application_controller.rb [hunk 10]
```diff
@@ -3,4 +3,11 @@ class ApplicationController < ActionController::Base
   # For APIs, you may want to use :null_session instead.
   protect_from_forgery with: :exception
   include SessionsHelper
+  
+  
+  # Confirms an admin employee.
+    def admin_employee
```
Comment (v-ramachandran) [inline app/controllers/application_controller.rb:9]: As a clarification point, will this always be the consistent behavior if the employee is not admin? If so, have it reflect in the method name itself (e.g. redirect_to_root_unless_admin).

File: app/views/sessions/new.html.erb (hunk 3 above)
Comment (tsmith-dev) [inline app/views/sessions/new.html.erb:14]: Would this be safe in the event that someone logged in on a public computer? Would you just assume that they would know to logout if they didn't want their session to be remembered or include some sort of message?

File: app/views/layouts/application.html.erb (hunk 4 above)
Comment (tsmith-dev) [inline app/views/layouts/application.html.erb:48]: Fixed in later task.

File: app/controllers/departments_controller.rb (hunk 8 above)
Comment (tsmith-dev) [inline app/controllers/departments_controller.rb:2]: Fixed in later task submission.

File: app/views/sessions/new.html.erb (hunk 3 above)
Comment (v-ramachandran) [inline app/views/sessions/new.html.erb:14]: It depends. While my original comment was not to force a change, it was more to point out that you may want to consider simplify the need to have the customer "check" the "Remember Me" button if the situation presents itself. The classic example is Amazon.com. If you sign-on, your session exists for quite a long time. Unless you actually need to manage your account, check your orders, etc. you don't have to log in again. An example of the convenience is that if prime customers want to go straight to watching videos, they don't have to log in every 30 or so minutes.

In general, while the resource you're going off of may mention the "Remember Me," you may want to consider if this can be simplified. For example, if you are making an in-office application, would that be used in a public computer? If not, then, is it still worthwhile? It might be or it might not, based on the privileges you allow. Overall, it's a design choice you may want to make based off your requirements and what you deem to be secure. Again, the point of the comment was more to push for considering the consequences of this, which in a real-world system you would decide during the design phase, rather than asking you to make a change.
